File: 0039/main.py
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if not candidates: return []
        
        candidates.sort()
        temp = [[x] for x in candidates]
        res = list()
        
        if target in candidates: res.append([target])
        for _ in range(target//candidates[0]):
            res_temp = temp.copy()
            temp = list()
            for L in res_temp:
                for x in candidates:
                    if x < L[-1]: continue
                    temp_L = L + [x]
                    sum_L = sum(temp_L)
                    if sum_L == target:
                        res.append(temp_L)
                    elif sum_L < target:
                        temp.append(temp_L)
                    else:
                        break
        return res
                

File: 0039/main.cpp
class Solution {
private:
    vector<vector<int>> res;
    void helper(vector<int>& candidates, int k, int target, vector<int> nums) {
        if(target == 0) {
            res.push_back(nums);
            return;
        }
        if(k==candidates.size()) return;

        while(target >= 0) {
            helper(candidates, k+1, target, nums);
            target -= candidates[k];
            nums.push_back(candidates[k]);
        }
    }

public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        helper(candidates, 0, target, {});
        return res;
    }
};